lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java

```diff
@@ -133,6 +148,12 @@ private void checkSecurity(Class clazz) throws ClassNotFoundException {
       return;
     }
 
+    for (String trustedClass : getTrustedClasses()) {
```

Comment:
Can we maybe perform this security check before loading the class? Instead of 
```
      Class clazz = ClassUtils.forName(getData().getClassLoader(), name);
      checkSecurity(clazz);
```
we can do
```
      checkSecurity(name);
      Class clazz = ClassUtils.forName(getData().getClassLoader(), name);
```
Eventually, we can tell the result of the security check by inspecting the name of the class, there's no need to load it with the ClassLoader no? Appears that a similar approach was taken in Parquet too.

Reply:
I agree that this would be slightly better but I don't see a problem with the class loading. As long as it is not instantiated it should be OK.

Reply:
Both is possible and should work basically the same. Ok to update the check.

Reply:
I will update with your suggestion

Reply:
By the way, if we do that the package check won't work. So if we are good to remove package check (and use class name only) that's fine.

Reply:
What's the problem with the package check ?

Reply:
> I agree that this would be slightly better but I don't see a problem with the class loading. As long as it is not instantiated it should be OK.

Although we block class instantiation if the given class isn't whitelisted, one can still trigger execution of a static initialiser block of any class from the classpath I think, when we load the class before checking if it is whitelisted or not. When we don't even load the class, then it isn't cause any possible headache either in the future.

Reply:
@martin-g to get the package, we need to get the class (with `ClassUtils.forName()`). If we do the `checkSecurity` with just class name (not the class itself), then we don't do package check.
If we want to do both, we need to split the check in two steps:
- class name first
- package once we have the class loaded

As we deprecate packages (in favor of class name), I think we can focus only on class name, right ?

Reply:
Why does the prefix matching logic for the package name implemented in [Parquet](https://github.com/apache/parquet-java/blob/master/parquet-avro/src/main/java/org/apache/parquet/avro/ReflectClassValidator.java#L57) not work here?

Reply:
Yes using the fqn

Reply:
Let me do that

Reply:
I did the requested change.